overlord/hookstate/hookmgr.go

```diff
@@ -0,0 +1,87 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package hookstate implements the manager and state aspects responsible for
+// the running of hooks.
+package hookstate
+
+import (
+	"fmt"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/hooks"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// DispatchHook is the hook dispatcher; exported here so it can be overridden in
+// tests.
+var DispatchHook = hooks.DispatchHook
+
+// HookManager is responsible for the maintenance of hooks in the system state.
+// It runs hooks when they're requested, assuming they're present in the given
+// snap. Otherwise they're skipped with no error.
+type HookManager struct {
+	state  *state.State
+	runner *state.TaskRunner
+}
+
+// Manager returns a new HookManager.
+func Manager(s *state.State) (*HookManager, error) {
+	runner := state.NewTaskRunner(s)
+	manager := &HookManager{
+		state:  s,
+		runner: runner,
+	}
+
+	runner.AddHandler("run-hook", manager.doRunHook, nil)
+
+	return manager, nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *HookManager) Ensure() error {
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait implements StateManager.Wait.
+func (m *HookManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop implements StateManager.Stop.
+func (m *HookManager) Stop() {
+	m.runner.Stop()
+}
+
+func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
+	task.State().Lock()
+	var hookRef hooks.HookRef
+	if err := task.Get("hook", &hookRef); err != nil {
+		return fmt.Errorf("failed to extract hook from task: %s", err)
```

Comment:
here you would be returning with the lock held, see other places, the pattern is to either Unlock before checking the error, or use what you had before (using defer) and unlock/lock around places where you don't need the lock (returning/panicing with the lock not held is fine so that's ok)